cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/operation/InMemoryOperationRunner.java

```diff
@@ -0,0 +1,145 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.Nullable;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link OperationRunner} to run an operation in the same service.
+ */
+public class InMemoryOperationRunner<T> extends AbstractOperationRunner<T> implements
+    AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryOperationRunner.class);
+
+  private final OperationStatePublisher statePublisher;
+  private final Lock lock;
+  private final ExecutorService executor;
+  private LongRunningOperation<T> currentOperation;
+
+  /**
+   * Default constructor.
+   */
+  @Inject
+  public InMemoryOperationRunner(Map<OperationType, LongRunningOperation> operations,
+      OperationStatePublisher statePublisher) {
+    super(operations);
+    this.statePublisher = statePublisher;
+    this.lock = new ReentrantLock();
+    this.executor = Executors.newSingleThreadExecutor(
+        Threads.createDaemonThreadFactory("operation-runner-%d")
+    );
+  }
+
+  // TODO(samik) It is currently rudimentary implementation to unblock the state management
+  //  Handle states, publish failures and heartbeat messages
+  @Override
+  public void run(LongRunningOperationRequest<T> request)
+      throws OperationTypeNotSupportedException, InvalidRequestTypeException {
+    lock.lock();
+    try {
+      if (currentOperation != null) {
+        throw new IllegalStateException("Operation already running");
+      }
+      currentOperation = getOperation(request);
+      executor.execute(() -> {
+        runInternal(request);
+      });
+      statePublisher.publishRunning();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void runInternal(LongRunningOperationRequest<T> request) {
+    try {
+      ListenableFuture<OperationError> operationTask = currentOperation.run(request.getOperationRequest(), this::sendMetaUpdateMessage);
+      Futures.addCallback(
+          operationTask,
+          new FutureCallback<OperationError>() {
+            @Override
+            public void onSuccess(@Nullable OperationError result) {
+              if (result != null) {
+                statePublisher.publishFailed(result);
+              } else {
+                statePublisher.publishSuccess();
+              }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+              statePublisher.publishFailed(getOperationErrorFromThrowable(request, t));
+            }
+          }
+      );
+    } catch (Exception e) {
+      statePublisher.publishFailed(getOperationErrorFromThrowable(request, e));
+    }
+  }
+
+  // TODO(samik): Implement graceful shutdown
+  @Override
+  public void stop(long timeout, TimeUnit timeoutUnit) throws Exception {
```

Comment:
Having both stop and close methods seems awkward. If you want a lifecycle support with graceful shutdown, then use `Service` instead of Closeable and have resource allocation happens in the start() method.